src/test/kotlin/me/yujinyan/retrofit/MoshiResultTypeAdapterFactory.kt

```diff
@@ -41,6 +40,7 @@ class MoshiResultTypeAdapterFactory : JsonAdapter.Factory {
           "errcode" -> errcode = reader.nextString().toIntOrNull()
           "msg" -> msg = reader.nextString()
           "data" -> data = dataTypeAdapter.fromJson(reader)
+            ?: throw JsonDataException("Expected non-null field [data], but was actually null")
```

Comment:
Not allowed. Since there is no way to get any information about whether the type variable is nullable or not, maybe it would be safer to avoid NPE by disallowing nullable types? This is still not a final solution and just for discussion.

Reply:
https://github.com/square/moshi/issues/1218#issuecomment-787772671